### ISSUE TITLE
iree-compile fails because of the lowering to builtin.unrealized_conversion_cast operations 

### ISSUE THREAD

**Issue by NicolaLancellotti** [iree-org/iree#15364]
### What happened?

During the compilation of the following reproducer, IREE fails with the error: `failed to legalize operation 'builtin.unrealized_conversion_cast' that was explicitly marked illegal`.

### Steps to reproduce your issue

model.mlir:
```
module @execute_model attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<2592000x3xf32>, %arg1: tensor<518400x1xi32>) -> tensor<518400xf32> {
      %0 = stablehlo.slice %arg0 [0:2592000, 0:1] : (tensor<2592000x3xf32>) -> tensor<2592000x1xf32>
      %1 = stablehlo.reshape %0 : (tensor<2592000x1xf32>) -> tensor<2592000xf32>
      %2 = "stablehlo.gather"(%1, %arg1) {dimension_numbers = #stablehlo.gather<collapsed_slice_dims = [0], start_index_map = [0], index_vector_dim = 1>, indices_are_sorted = false, slice_sizes = dense<1> : tensor<1xi64>} : (tensor<2592000xf32>, tensor<518400x1xi32>) -> tensor<518400xf32>
      return %2 : tensor<518400xf32>
  }
}
```

Command: `iree-compile --iree-input-type=stablehlo --iree-hal-target-backends=llvm-cpu model.mlir`

### Version information

commit 7b92a6d38599d3255b4513f1019c540a3401cfd3

**Comment by banach-space** [comment]
CC @jpienaar @rsuderman

**Comment by jpienaar** [comment]
Mmm that's weird. All f32s here ... @rsuderman and I'll take a look later here

**Comment by rsuderman** [comment]
Looks like there is a legalize missing from `vector.gather` to llvm. I believe ARM was involved with implementing the `vector.gather` work? @dcaballe do you happen to know much about ARMs vector gather work.

**Comment by rsuderman** [comment]
It appears the lowering from `vector.gather` to `llvm.gather` fails due to striding not being equal to 1. Is there a point in the pipeline when we can correct for this? Apologies as I don't know as much about the memref / striding behavior.

Simplified repro below:

```
// iree-opt --convert-to-llvm
#map = affine_map<()[s0] -> (s0 * 4096)>
#map1 = affine_map<()[s0] -> (s0 * -4096 + 518400, 4096)>
module {
  func.func @main_dispatch_0_generic_518400_i32xf32() {
    %cst = arith.constant dense<0.000000e+00> : vector<4xf32>
    %cst_0 = arith.constant dense<true> : vector<4xi1>
    %c0 = arith.constant 0 : index
    %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<2592000x3xf32>
    memref.assume_alignment %0, 64 : memref<2592000x3xf32>
    %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<518400xi32>
    memref.assume_alignment %1, 64 : memref<518400xi32>
    %2 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c0) : memref<518400xf32>
    memref.assume_alignment %2, 64 : memref<518400xf32>
    %subview = memref.subview %0[0, 0] [2592000, 1] [1, 1] : memref<2592000x3xf32> to memref<2592000xf32, strided<[3]>>
    %workgroup_id_x = hal.interface.workgroup.id[0] : index
    %3 = affine.apply #map()[%workgroup_id_x]
    %4 = affine.min #map1()[%workgroup_id_x]
    %subview_1 = memref.subview %2[%3] [%4] [1] : memref<518400xf32> to memref<?xf32, strided<[1], offset: ?>>
    %5 = vector.load %1[%3] : memref<518400xi32>, vector<4xi32>
    %6 = arith.index_cast %5 : vector<4xi32> to vector<4xindex>
    %7 = vector.gather %subview[%c0] [%6], %cst_0, %cst : memref<2592000xf32, strided<[3]>>, vector<4xindex>, vector<4xi1>, vector<4xf32> into vector<4xf32>
    vector.store %7, %subview_1[%c0] : memref<?xf32, strided<[1], offset: ?>>, vector<4xf32>
    return
  }
}
```

**Comment by banach-space** [comment]
Thanks for triaging this @rsuderman , that's super helpful!

>  I believe ARM was involved with implementing the vector.gather work? @dcaballe do you happen to know much about ARMs vector gather work.

You might be referring to the `tensor.extract` work that I did - let me assign it to myself. @dcaballe , please correct me if there's something else that I'm missing here.

**Comment by dcaballe** [comment]
Sorry, missed this. How is that we end up with that stride 3? Perhaps we could move the stride to the gather base/offset?

**Comment by banach-space** [comment]
> How is that we end up with that stride 3?

Not an StableHLO expert, but IIUC this is a strided slice:
```mlir
%0 = stablehlo.slice %arg0 [0:2592000, 0:1] : (tensor<2592000x3xf32>) -> tensor<2592000x1xf32>
```

> Perhaps we could move the stride to the gather base/offset?

Would that help? We'd still get a strided gather, right?

**EDIT:** Ignore the bit below - I got confused 🤦🏻 .
```mlir
%subview = memref.subview %0[0, 0] [2592000, 1] [1, 1] : memref<2592000x3xf32> to memref<2592000xf32, strided<[3]>>
// ...
%7 = vector.gather %subview[%c0] [%6], %cst_0, %cst : memref<2592000xf32, strided<[3]>>, vector<4xindex>, vector<4xi1>, vector<4xf32> into vector<4xf32>
```
TBH, I don't quite understand why both the `subview` and `gather` are strided.

**Comment by rsuderman** [comment]
Attaching the lowered linalg as it may provide some better clarity:
```
#map = affine_map<(d0, d1) -> (d0, d1)>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_embedded_elf_arm_64_]}>
module @execute_model attributes {hal.device.targets = [#device_target_llvm_cpu]} {
  func.func public @main(%arg0: tensor<2592000x3xf32>, %arg1: tensor<518400x1xi32>) -> tensor<518400xf32> {
    %extracted_slice = tensor.extract_slice %arg0[0, 0] [2592000, 1] [1, 1] : tensor<2592000x3xf32> to tensor<2592000x1xf32>
    %collapsed = tensor.collapse_shape %extracted_slice [[0, 1]] : tensor<2592000x1xf32> into tensor<2592000xf32>
    %0 = tensor.empty() : tensor<518400x1xf32>
    %1 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%arg1 : tensor<518400x1xi32>) outs(%0 : tensor<518400x1xf32>) {
    ^bb0(%in: i32, %out: f32):
      %2 = arith.index_cast %in : i32 to index
      %extracted = tensor.extract %collapsed[%2] : tensor<2592000xf32>
      linalg.yield %extracted : f32
    } -> tensor<518400x1xf32>
    %collapsed_0 = tensor.collapse_shape %1 [[0, 1]] : tensor<518400x1xf32> into tensor<518400xf32>
    return %collapsed_0 : tensor<518400xf32>
  }
}
```

**Comment by banach-space** [comment]
I've spent a bit more time looking at this and now I realise what @rsuderman and @dcaballe had in mind, sorry I missed that! 

IIUC, this could be fixed in two places:
1. in StableHLO lowerings so that no strided `memref.subview`s are generated (higher level),
2. in Vector canonicalisation so that strides from `memref.subview`s are replaced with modified gather indices (@dcaballe , is that what you had in mind?).

I'd rather have this fixed higher up the stack, but I'd need some guidance for that. Any suggestions?

Below is just an illustration of what Option 2. would entail.

 ### Before (original repro simplified, with strided `memref.subview`)
 
 **Does not lower to LLVM**

```mlir
#map = affine_map<()[s0] -> (s0 * 4096)>
#map1 = affine_map<()[s0] -> (s0 * -4096 + 518400, 4096)>
module {
  // Based on the original example
  func.func @example_1(%arg0 : memref<100x3xf32>, %arg0_1 : memref<100xf32>, %arg1: memref<518400xf32>, %idxs : vector<4xindex>, %x : index, %y : index) {
    %c0 = arith.constant 0 : index
    %x_1 = affine.apply #map()[%x]
    %subview = memref.subview %arg0[0, 0] [100, 1] [1, 1] : memref<100x3xf32> to memref<100xf32, strided<[3]>>
    %cst_0 = arith.constant dense<true> : vector<4xi1>
    %cst = arith.constant dense<0.000000e+00> : vector<4xf32>
    // The following line _does not_ get lowered to llvm.intr.masked.gather
    %7 = vector.gather %subview[%c0] [%idxs], %cst_0, %cst : memref<100xf32, strided<[3]>>, vector<4xindex>, vector<4xi1>, vector<4xf32> into vector<4xf32>
    // The following _does_ get lowered to llvm.intr.masked.gather
    //%7 = vector.gather %arg0_1[%c0] [%idxs], %cst_0, %cst : memref<100xf32>, vector<4xindex>, vector<4xi1>, vector<4xf32> into vector<4xf32>
    %subview_1 = memref.subview %arg1[%x_1] [%y] [1] : memref<518400xf32> to memref<?xf32, strided<[1], offset: ?>>
    vector.store %7, %subview_1[%c0] : memref<?xf32, strided<[1], offset: ?>>, vector<4xf32>
    return
  }
```

### After (strided `memref.subview` is replaced with index arithmetic)

**Does** lower to LLVM
```mlir
#map = affine_map<()[s0] -> (s0 * 4096)>
#map1 = affine_map<()[s0] -> (s0 * -4096 + 518400, 4096)>
  func.func @example_2(%arg0 : memref<100x3xf32>, %arg1: memref<518400xf32>, %idxs : vector<4xindex>, %x : index, %y : index) {
    %c0 = arith.constant 0 : index
    %x_1 = affine.apply #map()[%x]
    %cst_0 = arith.constant dense<true> : vector<4xi1>
    %cst = arith.constant dense<0.000000e+00> : vector<4xf32>
    %cst_3 = arith.constant dense<3> : vector<4xindex>
    %idxs_1 = arith.muli %cst_3, %idxs : vector<4xindex>
    %arg0_cast = memref.collapse_shape %arg0 [[0, 1]] : memref<100x3xf32> into memref<300xf32>
    %7 = vector.gather %arg0_cast[%c0] [%idxs_1], %cst_0, %cst : memref<300xf32>, vector<4xindex>, vector<4xi1>, vector<4xf32> into vector<4xf32>
    %subview_1 = memref.subview %arg1[%x_1] [%y] [1] : memref<518400xf32> to memref<?xf32, strided<[1], offset: ?>>
    vector.store %7, %subview_1[%c0] : memref<?xf32, strided<[1], offset: ?>>, vector<4xf32>
    return
  }
}
```

**Comment by banach-space** [comment]
Rob's latest post made me realise that this is an issue in the Linalg vectorization (or sth further down) rather than StableHLO lowerings, sorry. Definitely `tensor.extract` issue and definitely mea culpa. 

Thanks for the help narrowing this down!  I'll look into `tensor.extract` vectorization tomorrow.

**Comment by dcaballe** [comment]
If the fix shouldn't happen at StableHLO, I would approach this as gather canonicalization pattern and not as something the vectorizer should emit in one shot. So basically, we can add a canonicalization pattern to the gather op that turns:

```
    %subview = memref.subview %arg0[0, 0] [100, 1] [1, 1] : memref<100x3xf32> to memref<100xf32, strided<[3]>>
    %7 = vector.gather %subview[%c0] [%idxs], %cst_0, %cst : memref<100xf32, strided<[3]>>, vector<4xindex>, vector<4xi1>, vector<4xf32> into vector<4xf32>
```
into:
```
    %7 = vector.gather %arg0_cast[%c0] [%idxs_1], %cst_0, %cst : memref<300xf32>, vector<4xindex>, vector<4xi1>, vector<4xf32> into vector<4xf32>
```
WDYT?

I'm also curious about what would happen if instead of a gather we have a standard transfer read here... Gathers are currently only generated from `tensor.extract` ops so we wouldn't generate a gather for that "strided transfer read".

**Comment by banach-space** [comment]
@NicolaLancellotti Apologies that this is taking longer than expected. I've finally drafted a PR for MLIR:

* https://github.com/llvm/llvm-project/pull/72991

I'm not happy with how it is tested ATM, but it does fix the issue reported here and follows Diego's suggestion. Now, for you to be able to use that fix (without cherry-picking):
1. my MLIR change needs to land in llvm-project (_in progress_),
2. IREE's llvm-project fork needs to be updated (done fairly regularly anyway),
3. I need to update IREE to enable my new pattern (_todo_).

Hopefully the workaround that I shared with you offline is still sufficient and you are not blocked? 🤞🏻 :)

-Andrzej
